plugins/janus_audiobridge.c

```diff
@@ -1503,17 +1535,35 @@ void janus_audiobridge_incoming_rtp(janus_plugin_session *handle, int video, cha
 		}
 		/* Enqueue the decoded frame */
 		janus_mutex_lock(&participant->qmutex);
-		g_queue_push_tail(participant->inbuf, pkt);
-		/* Keep the queue at max of 150 packets, though */
-		if(g_queue_get_length(participant->inbuf) > 150) {
-			/* Remove the oldest packet */
-			janus_audiobridge_rtp_relay_packet *pkt = g_queue_pop_head(participant->inbuf);
-			if(pkt != NULL) {
-				if(pkt->data)
-					g_free(pkt->data);
-				pkt->data = NULL;
-				g_free(pkt);
-				pkt = NULL;
+		/* Insert packets sorting by sequence number */
+		participant->inbuf = g_list_insert_sorted(participant->inbuf, pkt, &janus_audiobridge_rtp_sort);
+		if(participant->prebuffering) {
+			/* Still pre-buffering: do we have enough packets now? */
+			if(g_list_length(participant->inbuf) == DEFAULT_PREBUFFERING) {
+				participant->prebuffering = FALSE;
+				JANUS_LOG(LOG_VERB, "Prebuffering done! Finally adding the user to the mix\n");
+			} else {
+				JANUS_LOG(LOG_VERB, "Still prebuffering (got %d packets), not adding the user to the mix yet\n", g_list_length(participant->inbuf));
+			}
+		} else {
+			/* Make sure we're not queueing too many packets: if so, get rid of the older ones */
+			if(g_list_length(participant->inbuf) >= DEFAULT_PREBUFFERING*2) {
```

Comment:
A dynamic value based on the connection quality would bring you back to step one. The main reason why we introduced this limited queue was to prevent the problem of users with troubled networks whose audio would increasingly become delayed over time, as we were queueing their audio indefinitely. Adapting the queue for those users and making it grow for them would cause the exact same problem over time.

Thinking about it, if you're talking about the receiving part (users getting audio) packets being dropped aren't necessarily being caused by this feature, and actually the cause may be substantially different. If users can't cope with the bitrate of the audio stream they're receiving because it's too high quality, a potential soultion would be to limit the bitrate for them only when we detect problems (e.g., encode a 16kHz stream for them although the room is 48kHz). We have an event that can be used this (plugins receive a slow_link event when too many NACKs are sent/received), so that can be used as a trigger to change the encoding parameters on the fly for the affected user. I'll have to check how easy it would be to add this to AudioBridge, but I'll only be able to look into that in a week more or less (busy in Japan ATM). The downside here is that we have no event for when things start being good again and you might increase the quality again, but it's a start.